---
title: "CiberAMP"
author:
- name: CIBERONC Bioinformatics workmodule
  email: ruben.fdez.caloto@gmail.com
package: ciberAMP
output:
  BiocStyle::html_document
abstract: |
    CiberAMP is a new R package to integrate somatic copy number and mRNA expression data from TCGA cohorts. It uses a new approach instead the Pearson correlation-based          previously used, that consists in two differential expression analysis (DEA): first, it calculates mRNA alterations between tumor and normal samples and, secondly, it         performs another DEA between copy number-altered and diploid tumor ones.\n
    To do so, CiberAMP takes advantage of GISTIC2.0 samples classification as diploid (0), shallowly amplified/deleted (+1/-1) or deeply amplified/deleted (+2/-2).\n
    Furthermore, this method also calculates how many well-knwon cancer-related genes may be co-amplified/deleted in the very same samples as our gene of interest.

vignette: |
  %\VignetteIndexEntry{CiberAMP}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
runtime: shiny
---

# Let's start!

The first step is to make sure you have installed all CiberAMP dependency packages:

* `r Biocpkg("TCGAbiolinks")`, for TCGA data downloading, normalizations and differential expression analysis (DEA).
* `r Biocpkg("SummarizedExperiment")`, for RNAseq counts matrices managing.
* `r Biocpkg("EDASeq")`, for RNAseq count data filtering and normalization.
* `r Biocpkg("edgeR")`, for mRNA differential expression analysis (DEA).
* `r Biocpkg("RTCGAToolbox")`, for TCGA cohorts last GISTIC2.0 Broad's Institute run thresholded data download.
* `r Biocpkg("dplyr")` & `r Biocpkg("stringr")`, for sample ID, and other strings, managing during the process.
* `r Biocpkg("ggplot2")` & `r Biocpkg("plotly")` & `r Biocpkg("shiny")`, for static (ggplot2) or interactive (plotly, shiny) result data plotting.

All these packages are used for TCGA data downloading and managing (normalization, differential expression analysis, copy number samples classification, etc...)

# Gene query

Once all packages are available, it is time to decide our genes of interest list. Many of you actually have a particular family of proteins or signaling pathway to consult. In that case, you should create a character vector with all the available and approved symbols for your queried genes, as in the example:
```{r, echo=TRUE}
genes.of.interest <- c("GENEA", "GENEB", "GENEC")
```
This vector will be passed to CiberAMP function at its "genes" argument.
In this example, we are going around a little bit more. In this paper, authors point out similaritis in 3p, 5p and 8p amplifications and 3q and 8q deletions in two squamous tumors cohorts from lung (LUSC) and head-and-nech (HNSC).
We are wonder if any of the genes lodged at these chromosomal arms would be significantly altered in its transcription due to such copy number alterations. Furthermore, we will connect these alterations to patients survival in order to reach out potential biomarkers with prognostic value in both cohorts.

# Getting genes from specific chromosomal locations

So, the first step was to create a data frame containing all the genes lodged at chromosome 3, 8 and chromosomal arm 5p. Actually, in this examples we are going to focus only on protein-coding genes, removing all the other type from the analysis.

```{r, echo=TRUE, width = 160}
library(biomaRt)
library(stringr)
library(GenomicRanges)

# First, we get all the genes inside 3p, 3q, 5p, 8p and 8q chromosomal arms.

ensembl <- useMart("ensembl")
ensembl <- useDataset("hsapiens_gene_ensembl",mart=ensembl)
all.genes <- getBM(attributes=c('chromosome_name', 'band', 'start_position', 'end_position', 'strand', 'ensembl_gene_id', 'hgnc_symbol', 'ensembl_transcript_id', 'gene_biotype', 'transcript_biotype'),
                   mart=ensembl)
head(all.genes)
all.genes$Chromosome.location <- paste0(all.genes$chromosome_name, all.genes$band)
all.genes <- all.genes[all.genes$gene_biotype %in% "protein_coding", ]
chr.regions <- c("3p", "3q", "5p", "8p", "8q")
for(i in 1:length(chr.regions)) {
  if(i == 1) {
    reg.genes <- all.genes[grep(chr.regions[i], all.genes$Chromosome.location), ]
  }else{
    reg.genes <- rbind(reg.genes, all.genes[grep(chr.regions[i], all.genes$Chromosome.location), ])
  }
}
reg.genes <- reg.genes[reg.genes$chromosome_name %in% c(3,5,8), ]
reg.genes <- reg.genes[!duplicated(reg.genes$ensembl_gene_id), ]
for(i in 1:nrow(reg.genes)) {
  if(reg.genes$strand[i] < 0) {
    reg.genes$strand[i] <- "-"
  }else{
    reg.genes$strand[i] <- "+"
  }
}
reg.genes <- reg.genes[reg.genes$hgnc_symbol != "", ]
head(reg.genes)
nrow(reg.genes)
```
As you can see, in reg.genes data frame are contained all the 1870 genes found in ENSEMBL mapped to the queried positions. We have information about where do they start, end, at which chromosome and strand. This data is crucial to easy map them into a chromose plot using `r Biocpkg("chromPlot")`.
Before that, `r Biocpkg("chromPlot")` requires a Genomics Range object as input. We can easily create one from reg.genes df just by using makeGRangesFromDataFrame function from `r Biocpkg("GenomicRanges")`:
```{r, echo=TRUE}
library(GenomicRanges)
gr <- makeGRangesFromDataFrame(reg.genes, seqnames.field = c("chromosome_name"), start.field = c("start_position"), end.field = c("end_position"), keep.extra.columns = TRUE, ignore.strand = FALSE)
gr
```

```{r all_in_genes, fig.cap="Queried genes. This plot shows these 1870 protein-coding genes chromosomal locations and density", fig.wide=TRUE, echo=TRUE}
library(chromPlot)
data(hg_gap)
chromPlot(gaps=hg_gap, annot1 = gr, chr = c(3,5,8))
```

# Running CiberAMP

CiberAMP will not use this mapping data. It is only necessary if we want to visualize our genes into the chromosomes as indicated in this example. We can skeep those steps if all we want is just running CiberAMP. This is the easiest thing ever: you only have to provide:

* A genes of interest character vector
* Those cohorts to be analyzed.
* The minimum % of copy number altered samples to be analyzed
* A path to a folder where results are going to be saved

```{r, echo=TRUE, results="hide"}
library(ciberAMP)
#x <- ciberAMP(genes = reg.genes$hgnc_symbol, cohorts = c("LUSC", "HNSC"), pat.percentage = 0, writePath = "/Users/lab02/Desktop/CiberAMP vignette folder")
x <- readRDS("/Users/lab02/Desktop/CiberAMP vignette folder/my_data.rds")
```
This is a computing expensive step, so we already provide you the results for the previous line. You can extract them easily.
```{r, echo=TRUE, results="hide"}
library(ciberAMP)
x <- readRDS("/Users/lab02/Desktop/CiberAMP vignette folder/my_data.rds")
```
In any case, if you have time to run it by yourself, we leave you here a description of all CiberAMP arguments for you notice:

* genes Your genes of interest official symbols. This should be a character vector.
* cohorts By default, CiberAMP will compute every TCGA cohort (the 32 available), but you can use this argument to only analyze a few. You can consult the offical codes at TCGA's web -> https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/tcga-study-abbreviations
* writePath The folder path to save the results. TIP: if you want to re-run CiberAMP, if you use the same folder where all data was stored, you will never need to re-download it again what saves a lot of time and disk space! But, be careful, every run your results will be overwritten!
* pat.percentage The minimum % of altered patients that should be considered to look for SCNA-driven differential expression. When in a cohort an event occurs recurrently it is more likely that such event may be under positive selection so being relevant for tumor fitness.
* pp.cor.cut Threshold to filter samples by AICC. Passed to `TCGAanalyze_Preprocessing`.
* norm.method Method of normalization, such as `gcContent` or `geneLength` (default).
* filt.method Method of filtering, such as `quantile` (default), `varFilter`, `filter1`, `filter2`.
* filt.qnt.cut Threshold selected as mean for filtering. Defaults to 0.25.
* filt.var.func Filtering function. Defaults to `IQR`. See `genefilter` documentation for available methods.
* filt.var.cutoff Threshold for `filt.var.funct`.
* filt.eta Parameter for `filter1`. Defaults to 0.05.
* filt.FDR.DEA Threshold to filter differentially expressed genes according their corrected p-value.
* filt.FC Parameter for `filter2`. Defaults to 1.
* p.val.thr Threshold for reported p values. Defaults to 1 (report all).
* cna.thr Threshold level for copy-number variation analysis. Can be `Deep`, `Shallow` or `Both`
* exp.mat Custom expression matrix, which will not be filtered. Defaults to `NULL`,
* cna.mat Custom copy-number analysis matrix, which will not be filtered. Defaults to `NULL`,

# Looking into CiberAMP results

CiberAMP results into a list of 3 data frames that can be accessed by:

The x[[1]] data frame contains differential expression results for queried genes:

* Column 1 -> queried gene approved symbol.
* Columns 2:4 -> queried genes tumor vs normal differential expression results.
* Column 5 -> TCGA cohort.
* Column 6:9 -> queried gene copy number altered vs diploid tumor samples differential expression results.
* Column 10  -> TCGA cohort.
* Column 11  -> CN-altered vs diploid tumors comparison type: amplified vs diploid or deleted vs diploid.
* Column 12  -> % of samples CN-altered.
* Column 14 -> CN-altered TCGA sample barcodes.

```{r queried genes results data frame, fig.cap="SCNA-driven differentially expressed genes.", fig.wide=TRUE, echo=TRUE}
head(x[[1]][1, ])
```

The x[[2]] data frame contains differential expression results for COSMIC cancer census genes list. Column information is exactly the same as x[[1]].

```{r COSMIC CGC genes results data frame, fig.cap="SCNA-driven differentially expressed genes.", fig.wide=TRUE, echo=TRUE}
head(x[[2]][1, ])
```

Finally, the x[[3]] contains overlapping information between queried and CGC genes copy number overlap between samples. Genes may be CN-altered due to being lodged next to well-known and transcriptionally active in cancer cancer-related genes. In this data frame, you can consult if any of your queried CN-alterations would be highly correlated to any of known oncodrivers or tumor suppressors:

* Column 1 -> queried gene approved symbol.
* Column 2:5 -> queried gene copy number altered vs diploid tumor samples differential expression results.
* Column 6 -> TCGA cohort.
* Column 7 -> CN-altered vs diploid tumors comparison type: amplified vs diploid or deleted vs diploid.
* Column 8 -> % of samples CN-altered in queried gene.
* Column 9 -> CN-altered TCGA sample barcodes in queried gene.
* Column 10 -> COSMIC CGC gene approved symbol.
* Column 11:14 -> COSMIC CGC gene copy number altered vs diploid tumor samples differential expression results.
* Column 15 -> TCGA cohort.
* Column 16 -> CN-altered vs diploid tumors comparison type: amplified vs diploid or deleted vs diploid.
* Column 17 -> % of samples CN-altered in COSMIC CGC gene.
* Column 18 -> CN-altered TCGA sample barcodes in COSMIC CGC gene.
* Column 19 -> % of queried gene CN-altered samples that overlaps with COSMIC CGC gene CN-altered samples.
* Column 20 -> % of COSMIC CGC gene CN-altered samples that overlaps with queried gene CN-altered samples.

```{r Overlap between query and COSMIC CGC genes CN-altered samples, fig.cap="SCNA-driven differentially expressed genes.", fig.wide=TRUE, echo=TRUE}
head(x[[3]][1, ])
```


# Mapping results into chromosomes

From the original 1870 genes we found that:

* 114 CN-altered queried genes show significant CN-driven differential expression in TCGA-HNSC cohort.
* 201 CN-altered queried genes show significant CN-driven differential expression in TCGA-LUSC cohort.
* Only 67 of them were found in both cohorts.

These are mapped in the next figure at their chromosomal positions:
```{r only_assoc_both_genes, fig.cap="SCNA-driven differentially expressed genes.", fig.wide=TRUE, echo=TRUE}
library(chromPlot)
df <- x[[1]]
genes_cosmic <- x[[3]]
gr.assoc.hnsc <- gr[gr$hgnc_symbol %in% as.character(df[df$Tumor %in% "HNSC" & abs(df$log2FC.SCNAvsDip) >= 1 & df$FDR.SCNAvsDip <= 0.01, ]$Gene_Symbol), ]
gr.assoc.lusc <- gr[gr$hgnc_symbol %in% as.character(df[df$Tumor %in% "LUSC" & abs(df$log2FC.SCNAvsDip) >= 1 & df$FDR.SCNAvsDip <= 0.01, ]$Gene_Symbol), ]
both <- intersect(gr.assoc.hnsc$hgnc_symbol, gr.assoc.lusc$hgnc_symbol)
chromPlot(gaps=hg_gap, annot1 = gr[gr$hgnc_symbol %in% both, ], chr = c(3,5,8))
```

The next question was: are these genes mapping very next to known COSMIC CGC ones? For that reason, we decided to graphically show it doing:

```{r genes_and_cosmic, fig.cap="SCNA-driven differentially expressed genes.", fig.wide=TRUE, echo=TRUE}
library(chromPlot)
df.cosmic <- x[[2]]
gr.assoc.hnsc.cosmic <- gr[gr$hgnc_symbol %in% as.character(df.cosmic[df.cosmic$Tumor %in% "HNSC" & abs(df.cosmic$log2FC.SCNAvsDip) >= 1 & df.cosmic$FDR.SCNAvsDip <= 0.01, ]$Gene_Symbol), ]
gr.assoc.lusc.cosmic <- gr[gr$hgnc_symbol %in% as.character(df.cosmic[df.cosmic$Tumor %in% "LUSC" & abs(df.cosmic$log2FC.SCNAvsDip) >= 1 & df.cosmic$FDR.SCNAvsDip <= 0.01, ]$Gene_Symbol), ]
both.cosmic <- intersect(gr.assoc.hnsc.cosmic$hgnc_symbol, gr.assoc.lusc.cosmic$hgnc_symbol)
chromPlot(gaps=hg_gap, annot1 = gr[gr$hgnc_symbol %in% both, ], annot2 = gr[gr$hgnc_symbol %in% both.cosmic, ], chr = c(3,5,8), chrSide=c(-1,1,1,1,1,1,1,1))
```

In this figure, rigth red bars show those 67 CN-driven diff. expressed queried genes and left yellow bars show 8 CN-driven differentially expressed COSMIC CGC genes at their genomic positions.

# Ggplotting the results

CiberAMP package actually gives you a ggplot function to plot your results. This graph cosists in: 1) a X axis representing log2(FC) values for each genes comparing tumor and normal samples mRNA differential expression, and 2) a Y axis representing log2(FC) values for each gene comparing copy number altered vs diploid tumor samples mRNA differential expression.
This plot will allow you to find out which or your genes are:

* Differentially expressed in tumor samples
* Differentially expressed in copy number altered samples.

And how these two data are connected:

* Upregulated genes which overexpression bursts due to copy number alteration -> x-value > 0; y-value > 0
* Upregualted genes exclusively when there is a copy alteration -> x-value = 0; y-value > 0
* Downregulated genes which copy number alteration increases very significantly its expression -> x-value < 0; y-value > 0
* Upregulated genes wich copy number alterations does not lead significant transcriptional alteration -> x-value != 0; y-value == 0
* Downregulated genes in tumors which copy number alteration even intensifies its repression -> x-value < 0; y-value < 0
* Downregulated genes exclusively repressed by copy number alteration -> x-value = 0, y-value < 0
* Upregualted genes which copy number alteration significantly represses its expression -> x-value > 0; y-value < 0

Here it is those 67 common CN-driven diff. expressed genes:

```{r ggplot, fig.cap="SCNA-driven differentially expressed genes.", fig.wide=TRUE, echo=TRUE}
library(ggplot2)
df.exp <- x[[1]]
df.exp <- df.exp[df.exp$Gene_Symbol %in% both, ]
ggplot.CiberAMP(df.exp)
```

# Interacting with the results

Finally, CiberAMP provides a function to interactively explore your results. This is an application that can be used to filter queried genes by:

* Gene symbol
* Minimum % of CN-altered samples
* log2(FC) threshold for tumor vs normal samples differential expression
* log2(FC) threshold for CN-altered vs diploid tumor samples differential expression.
* TCGA cohorts to be displayed (directly on graph's legend).

Actually, clicking on any point would display a table to show the overlapping samples % between queried and COSMIC CGC genes.

Tip: Click on submit button!!

```{r interactive.plot, fig.cap="SCNA-driven differentially expressed genes.", fig.wide=TRUE, echo=TRUE}
library(shiny)
library(plotly)
library(DT)
int.plot.CiberAMP(df.exp, x[[3]])
```

# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
