---
title: "CiberAMP"
author:
- name: CIBERONC Bioinformatics workmodule
  email: ruben.fdez.caloto@gmail.com
package: ciberAMP
output:
  BiocStyle::html_document
abstract: |
  CiberAMP is a new R package to integrate somatic copy number and mRNA expression data from TCGA cohorts. It uses a new approach instead the Pearson correlation-based previously used, that consists in two differential expression analysis (DEA): first, it calculates mRNA alterations between tumor and normal samples and, secondly, it performs another DEA between copy number-altered and diploid tumor ones.\n
  To do so, CiberAMP takes advantage of GISTIC2.0 samples classification as diploid (0), shallowly amplified/deleted (+1/-1) or deeply amplified/deleted (+2/-2).\n
  Furthermore, this method also calculates how many well-knwon cancer-related genes may be co-amplified/deleted in the very same samples as our gene of interest.
  
vignette: |
  %\VignetteIndexEntry{CiberAMP}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Let's start!

The first step is to make sure you have installed all the packages of which CiberAMP depends on. These are dedicated to data downloading and manage:

* `r Biocpkg("TCGAbiolinks")`, for TCGA data downloading, normalizations and differential expression analysis (DEA).
* `r Biocpkg("SummarizedExperiment")`, for RNAseq counts matrices managing.
* `r Biocpkg("EDASeq")`, for RNAseq count data filtering and normalization.
* `r Biocpkg("edgeR")`, for mRNA differential expression analysis (DEA).
* `r Biocpkg("RTCGAToolbox")`, for TCGA cohorts last GISTIC2.0 Broad's Institute run thresholded data download.
* `r Biocpkg("dplyr")` & `r Biocpkg("stringr")`, for sample ID, and other strings, managing during the process.
* `r Biocpkg("ggplot2")` & `r Biocpkg("plotly")` & `r Biocpkg("shiny")`, for static (ggplot2) or interactive (plotly, shiny) result data plotting.

# Gene query

Now we have to define our genes of interest list. In this vignette, we use CiberAMP to find out how characteristic squamous tumors copy gains at   and losses at   may drive transcriptional alteration in genes lodged at these regions.

First, we download human whole genome gene symbols list. Then, we filter it out and get only those lodged at the mentioned regions.

```{r, echo=TRUE}
library(biomaRt)
library(stringr)

# We download all human genome symbols using HGNC Biomart tool and read in "all_genes" variable.
all_genes <- read.table("/Users/lab02/Desktop/results.txt", sep="\t", header=TRUE, quote = "", fill = TRUE)
head(all_genes[c(3,6)])

# We specify which regions we are interested in and we "grep" all those genes which "chromosome.location" column items starts by any of the specified.
chr.regions <- c("3p", "3q", "5p", "8p", "8q")
for(i in 1:length(chr.regions)) {
  if(i == 1) {
    reg.genes <- all_genes[grep(chr.regions[i], all_genes$Chromosome.location), ]
  }else{
    reg.genes <- rbind(reg.genes, all_genes[grep(chr.regions[i], all_genes$Chromosome.location), ])
  }
}

# Here it is the filtered data frame...
head(as.data.frame(table(as.character(reg.genes$Chromosome.location))))

# ... and here our genes of interest vector.
genes.of.int <- as.character(reg.genes$Approved.symbol)
genes.of.int[1:4]
length(genes.of.int)
```

There are 6240 different genes lodged at the queried chromosomal arms. Now, CiberAMP is directly applied:

```{r, echo=TRUE}
library(ciberAMP)

x <- ciberAMP(genes = genes.of.int, cohorts = c("LUSC", "HNSC"), pat.percentage = 0, writePath = "/Users/lab02/Desktop/CiberAMP vignette folder")

```

This step is going to take some time... So I propose you to further read the discussion about CiberAMP function arguments. You will find some useful tips even.

* genes List of official gene symbols for analysis. See function \code{all_genes} for a list of available genes.
* cohorts List of tumors to study. See function \code{all_tumors} for a list of available tumors.
* writePath Path where results are stored. Defaults to the current folder.
* pat.percentage Minimum percentage of patients per group.
* pp.cor.cut Threshold to filter samples by AICC. Passed to `TCGAanalyze_Preprocessing`.
* norm.method Method of normalization, such as `gcContent` or `geneLength` (default).
* filt.method Method of filtering, such as `quantile` (default), `varFilter`, `filter1`, `filter2`.
* filt.qnt.cut Threshold selected as mean for filtering. Defaults to 0.25.
* filt.var.func Filtering function. Defaults to `IQR`. See `genefilter` documentation for available methods.
* filt.var.cutoff Threshold for `filt.var.funct`.
* filt.eta Parameter for `filter1`. Defaults to 0.05.
* filt.FDR.DEA Threshold to filter differentially expressed genes according their corrected p-value.
* filt.FC Parameter for `filter2`. Defaults to 1.
* p.val.thr Threshold for reported p values. Defaults to 1 (report all).
* cna.thr Threshold level for copy-number variation analysis. Can be `Deep`, `Shallow` or `Both`
* exp.mat Custom expression matrix, which will not be filtered. Defaults to `NULL`,
* cna.mat Custom copy-number analysis matrix, which will not be filtered. Defaults to `NULL`,



#' @param p.val.thr Threshold for reported p values. Defaults to 1 (report all).
#' @param cna.thr Threshold level for copy-number variation analysis. Can be `Deep`, `Shallow` or `Both`.
#' `Deep`, consider likely homozygous deletions for loss and high-level amplifications (high copy-number, focal) for gain.
#' `Shallow`, consider likely heterozygous deletions for loss and low-level amplifications (lower copy-number, broad) for gain.
#' `Both`, consider all loss and gain events.
#' @param exp.mat Custom expression matrix, which will not be filtered. Defaults to `NULL`,
#' which means that the method considers TCGA expression matrices belonging to `cohorts`.
#' @param cna.mat Custom copy-number analysis matrix, which will not be filtered. Defaults to `NULL`,
#' which means that the method considers TCGA copy-number analysis matrices belonging to `cohorts`.

Small and wide figures can be specified by `fig.small` and `fig.wide` code chunk options.

```{r small, fig.cap="Small figure. A plot produced by a code chunk with option `fig.small = TRUE`.", fig.small=TRUE, echo=FALSE}
plot(cars)
```

```{r wide, fig.cap="Wide figure. A plot produced by a code chunk with option `fig.wide = TRUE`.", fig.wide=TRUE, echo=FALSE}
plot(cars)
```


# Equations

To number and reference equations, put them in equation environments and assign labels to them, see Equation \@ref(eq:binom).

\begin{equation}
  f\left(k\right) = \binom{n}{k} p^k\left(1-p\right)^{n-k}
  (\#eq:binom)
\end{equation}


# Tables

Like figures, tables with captions will also be numbered and can be referenced, see Table \@ref(tab:table).

Fruit   | Price
------- | -----
bananas | 1.2
apples  | 1.0
oranges | 2.5

: (\#tab:table) A simple table. With caption.


# Cross-references

Apart from referencing figures (Section \@ref(figures)), tables (Section \@ref(tables)), and equations (Section \@ref(equations)), you can also use the same syntax to refer to sections by their default labels generated by pandoc.


# Side notes

Footnotes are displayed as side notes on the right margin^[this is a side note entered as a footnote], which has the advantage that they appear close to the place where they are defined.


# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
